pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -113,15 +113,19 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(env model.Environmen
 			// this is for.
 			FilterChains: make([]plugin.FilterChain, len(l.FilterChains)),
 		}
-		for _, p := range configgen.Plugins {
-			params := &plugin.InputParams{
-				ListenerType:   listenerType,
-				Env:            &env,
-				Node:           &node,
-				ProxyInstances: workloadInstances,
-			}
-			if err = p.OnOutboundListener(params, mutable); err != nil {
-				log.Warna("buildGatewayListeners: failed to build listener for gateway: ", err.Error())
+
+		for _, w := range workloadInstances {
```

Comment:
I don't think this is correct to iterate over all instances (which basically one for each port of ingress service). Listener is created for gateway's `portNumber` (see line 94), so the `ServiceInstance` here should be a combine of (ingressgateway) service hostname and this port. This could be tricky to generate that (you can forge it, but not sure if it safe to do so)

On the other hand, I don't know if gateway port can be only one of the ports that ingressgateway service provides. In that case, may be we can just select the workload that matching such port. + @rshriram

Reply:
From debug logs of [integration test](https://github.com/istio/istio/blob/master/tests/e2e/tests/pilot/ingressgateway_test.go#L44) which applies [ingressgateway.yaml](https://github.com/istio/istio/blob/master/tests/e2e/tests/pilot/testdata/v1alpha3/ingressgateway.yaml), two listeners on http port 80/443 are built(another tcp Port:31400), route_config of the 80 listener is like below
`
  "routes": [
                     {
                        "decorator": {
                           "operation": "gateway-bound-route"
                          },
                        "match": {
                           "prefix": "/"
                          },
                        "per_filter_config": {
                          },
                        "route": {
                           "cluster": "outbound|80|v2|c.istio-system.svc.cluster.local",
                           "use_websocket": false
                          }
                       }
                    ]
                 }
`

which means we may need to add filters on each port.

Reply:
Right, but you already iterate for that ports (loop start at line 74)

Reply:
thanks ! you're right :+1: